### ISSUE TITLE
Views Filters containing operator 'is none of' broken since 1.12.5

### ISSUE THREAD

**Issue by JDSaward** [backdrop/backdrop-issues#3614]
### Steps To Reproduce

Set up content:

1. Add a "status" field of type "list: text" to a post
2. Set the allowed values to applied, approved, and concluded
3. Add content, one post with each value

Set up view:

1. Create a test view. format: table, fields: title, status
2. Add filters: published = yes, type = post
3. Test the view, see all 3 posts

Add problem filter:

1. Add a filter for field status not equal to "concluded"
2. Test the view, see no results at all.

### Actual behavior

status not equal to "concluded" returns no results

### Expected behavior

status not equal to "concluded" should return 2 results

---
### Original report by @JDSaward 

After updating to 1.12.5 a view on user profiles previously delivering correct results was delivering an empty results set. 

I tracked it down to a views filter on the user roles: 'roles is none of tester'. 

Removing that filter delivered the result set (but including the users with the tester role, which are required to not display).

I suppose the issue arose in 1.12.4, as that release dealt with Views security issues.

I investigated further and the same issue has been logged for Drupal 7.x-3.21, marked as 'Priority Major' - see https://www.drupal.org/project/views/issues/3040391.  Currently (23 March 2018 1am) that issue has not yet been resolved for Drupal.

edit: adding some more search terms so I can find this issue again, next time I search for it...

**Views "not equal to" filter <> returns no results.**

---

Previous PRs:
~TESTS ONLY PR: https://github.com/backdrop/backdrop/pull/2877~ (fails as expected)
~PR by @jenlampton: https://github.com/backdrop/backdrop/pull/2582~ (this one Drupal/Views committed, but does not fix the problem)
~PR by @jenlampton: https://github.com/backdrop/backdrop/pull/2581~ (this one fixes the problem, was previously recommended in the Drupal/Views queue)
~PR by @jenlampton: https://github.com/backdrop/backdrop/pull/2870~ (this one includes the patch from https://www.drupal.org/node/2036473 - tests fail)
~~PR by @docwilmot: https://github.com/backdrop/backdrop/pull/2768~~ (weird test failure issues, only sometimes)

---

This final PR comments out the mysterious failing tests, but leaves them in with @todo comments. This makes the patch committable (to fix the bug), but leaves us the opportunity to [uncomment the tests and get them passing](https://github.com/backdrop/backdrop-issues/issues/4092) -- perhaps after all [the other views patches](https://github.com/backdrop/backdrop-issues/issues/3685) are committed.

**Final PR https://github.com/backdrop/backdrop/pull/2878**

**Comment by herbdool** [comment]
Thanks for reporting @JDSaward. Looks like they're close in the D7 issue, might just needs tests.

**Comment by klonos** [comment]
Thanks for taking the time to report this @JDSaward and welcome to the community!! 👋 ...I see that you are also Melbourne-based, which is awesome 👍 ...I have been waiting so long for Backdropers to show up in VIC, so we can start meetups 😄 🎉 ❤️ ...I'd encourage you to join our [Gitter channel](https://gitter.im/backdrop/backdrop-issues), since there's quite a few of us there lurking most of the time, so might be a better way to get quick responses for any question you may have.

Our [1.12.6 milestone](https://github.com/backdrop/backdrop-issues/milestone/72) has 2 bug fixes merged already, one of which is quite serious (#3519). We have another 5 bug fixes RTBCed in the same milestone, out of witch another 2 are also serious (#3249 and #3573).

So basically yeah, once this is fixed, I think that we'd probably qualify for a new bug fix release (plus we'll most likely have a few other issues RTBCed by then).

**Comment by laryn** [comment]
I can confirm that [this patch](https://www.drupal.org/files/issues/2019-03-25/views-not-operator-3040391-25.patch) (manually applied) is working for me. I'll add a PR on Backdrop's code in case others want to see the changes here and test it out.

**Comment by klonos** [comment]
...another person has reported that they are getting this error with that patch:

> Warning: Illegal offset type in views_many_to_one_helper->ensure_my_table() (line 1029 of views/includes/handlers.inc).

I think that it is best to help test the d.org patch, and then use what officially gets in Views 7.x

**Comment by laryn** [comment]
This is now marked RTBC on Drupal side: 
https://www.drupal.org/project/views/issues/3040391

**Comment by klonos** [comment]
Noting that the following 2 issues have been mentioned as follow-ups:

https://www.drupal.org/project/views/issues/1151164
https://www.drupal.org/project/views/issues/3040090

**Comment by jenlampton** [comment]
First PR implements fix recommended by @laryn 23 days ago:
https://github.com/backdrop/backdrop/pull/2581

Second PR implements fix committed to Drupal views module:
https://github.com/backdrop/backdrop/pull/2582

**Comment by klonos** [comment]
FYI, there are comments in the d.org issue after the [commit that was meant to fix this for Views 7.x](https://www.drupal.org/project/views/issues/3040391#comment-13052907), by some people saying that the committed change has not fixed things for them.

[This comment](https://www.drupal.org/project/views/issues/3040391#comment-13072910) specifically also mentions that the only patch that fixed things is the one @laryn has linked too:

>I've tried both patch applications mentioned by FeyP in #55 and neither one fixes my issue.
>
My issue still seems to stem from the 'not' operator used in a filter with the "Is none of" selected, with a single Option selected.
>Patch #25 is still the only patch that fixes my issue.

Some mention that there have been other related fixes committed that might have contributed to fixing the problems for the rest of the people. Perhaps we are missing something.

**Comment by klonos** [comment]
> Perhaps we are missing something.

Here's a start: #3685 😉 ...I will need 👀 for CR and confirm where things are N/A to us.

**Comment by jenlampton** [comment]
Since the "fix" Drupal committed does not fix the issue, I recommend we pursue the fix that works. Here's that PR again, for review: backdrop/backdrop#2581

**Comment by klonos** [comment]
Could this be it?: https://github.com/backdrop/backdrop/pull/2666 (or related somehow?)

**Comment by jenlampton** [comment]
@klonos the unresolved Drupal issue number is 3040391. That one looks more specific.

**Comment by jenlampton** [comment]
@matthui or @isabelhaas are either of you up for testing this one? I would love it if it could get in in the next release so I don't have to patch all my sites again.

**Comment by isabelhaas** [comment]
I tested the PR and it worked for me.

**Comment by Graham-72** [comment]
I have applied the PR manually to two sites that needed it and it works for these.

**Comment by jenlampton** [comment]
Thanks @Graham-72 and @isabelhaas I have added the `pr - works for me` label to indicate that this fix works for both of you.

**Comment by quicksketch** [comment]
Although the port from Drupal (https://github.com/backdrop/backdrop/pull/2582) may not fix the same problem in Backdrop, it has a comprehensive test of the broken functionality. We should pull out that test and included it in the working PR (https://github.com/backdrop/backdrop/pull/2581).

**Comment by laryn** [comment]
PR works for me as well.

**Comment by klonos** [comment]
It works for me too, but I just don't want us to accidentally revert this and break things again, in the process of getting the the Views 7.x crossports done (#3685).

Is there a reliable way to do that? ...inline `don't touch this! @see https://github.com/backdrop/backdrop-issues/issues/3614` comments perhaps?

**Comment by jenlampton** [comment]
@klonos maybe add a note to the commit in the views checklist and reference this issue.

**Comment by klonos** [comment]
@jenlampton I fear that it might be missed in the checklist. It won't if it's right there in the code.

**Comment by jenlampton** [comment]
@klonos It would be better not to add notes into the code for our process. Notes in the code should relate to the code :)

**Comment by klonos** [comment]
![Screen Shot 2019-06-06 at 4 26 26 am](https://user-images.githubusercontent.com/2423362/58980290-4bdc5d80-8813-11e9-9982-7ced839d8307.png)

**Comment by jenlampton** [comment]
Almost...  more like:
<img width="768" alt="Screen Shot 2019-06-05 at 11 30 59 AM" src="https://user-images.githubusercontent.com/397895/58980532-6cf66d00-8785-11e9-856e-78e21ec30cf0.png">

**Comment by klonos** [comment]
Well, not exaclty @jenlampton ...

[That commit](manytoone.test) introduces a `manytoone` test. If we mark it as done, we'd be missing that test. Are you thinking to include it in this PR here?

**Comment by jenlampton** [comment]
Yes, that's why this issue is marked "needs - automated tests".

**Comment by klonos** [comment]
I have filed https://github.com/backdrop/backdrop/pull/2727 (as part of #3853 - not for this issue here), so not to miss on the rest of the changes that went in Views 7.x as part of https://git.drupalcode.org/project/views/commit/9aaf2f8d1557438023ecdadad0f9f41d43d6f641 ...we are safer here now I think.

**Comment by klonos** [comment]
...so more like:

![Screen Shot 2019-06-06 at 8 47 10 am](https://user-images.githubusercontent.com/2423362/58995471-b56e6300-8837-11e9-9308-1310e19691ef.png)

(if you agree that is)

**Comment by jenlampton** [comment]
I've pushed an update to [this PR](https://github.com/backdrop/backdrop/pull/2582/files) that includes a test from [the other PR](https://github.com/backdrop/backdrop/pull/2582/files) but I haven't yet gotten it working. Put at least it's a start :)

**Comment by jenlampton** [comment]
Okay, so the some of the faulures in the test are due to our default content. The test compares user IDs to node IDs assuming there are no node IDs to start with.

I tried to switch to the testing profile to get rid of our default content, but everything else in this test depends on so much of the other stuff in standard, I'm thinking this isn't a good route to take. 

Next I plan to create two dummy users to match the default content IDs, and see if I can get the numbers to match up that way.

**Comment by jenlampton** [comment]
Okay, I got most of the tests passing by cheating and creating three users. I've still got a 3 exceptions, and a few tests I couldn't fix. I'm out of steam. Unasigning.

**Comment by jenlampton** [comment]
This issue is fixed, it needs to be committed to core. I've now spent more than 7 hours working on a stupid test. There's no way this test is this important. Let's just fix the bug and move on.

**Comment by serundeputy** [comment]
Can we remove the default content now that users upon install are directed to the dashboard? Would that make the tests easier/doable? The default content was intended to welcome and give novice users an idea of what to do; I think the dashboard does that now.

**Comment by jenlampton** [comment]
> Can we remove the default content now that users upon install are directed to the dashboard?

I don't think the dashboard trains people how to edit content. But that Q should be in a new issue if you are serious about it :)

**Comment by klonos** [comment]
> The default content was intended to welcome and give novice users an idea of what to do; I think the dashboard does that now.

Yup, but that happens in the standard profile. Other profiles that do not enable the Dashboard module by default could still use the default content.

**Comment by jenlampton** [comment]
We still need default content to edit. Dashboard has nothing to do with
node edit forms.

**Comment by docwilmot** [comment]
https://github.com/backdrop/backdrop-issues/issues/3614#issuecomment-484319842

> First PR implements fix recommended by @laryn 23 days ago:
> [backdrop/backdrop#2581](https://github.com/backdrop/backdrop/pull/2581)
> This one **works** for me.
> 
> Second PR implements fix committed to Drupal views module:
> [backdrop/backdrop#2582](https://github.com/backdrop/backdrop/pull/2582)
> This one **does not work** for me.

Just created a View with the patch from [backdrop/backdrop#2754](https://github.com/backdrop/backdrop/pull/2754) (which replaced 2581 and it does NOT work for me. I think that may be the reason the tests dont pass.

View:
- new content type foo
- list field 'mystatus' options good, bad, ugly
- filter view by content type and by 'mystatus'
- if you filter by NOT but only select one option (good bad OR ugly) it works
- if you filter by multiple NOTs (not good nor bad, for example) it doesnt.

**Comment by docwilmot** [comment]
Used the fix that was committed to D7. All passing local but one test fail on Zen CI. Not sure why. But would be good if someone could test this locally please?

**Comment by docwilmot** [comment]
Thought I'd try the original test from the D7 patch and surprise, it passes on Zen, but fails on my local.

**Comment by klonos** [comment]
Hmm 🤔 php/webserver/mysql versions on your local @docwilmot ??

**Comment by klonos** [comment]
I like keeping parity with what's been committed in Views 7.x, so 👍 @docwilmot 

I have filed a PR with just the tests from your latest PR. I expect things to fail: https://github.com/backdrop/backdrop/pull/2769

**Comment by klonos** [comment]
>Thought I'd try the original test from the D7 patch and surprise, it passes on Zen, but fails on my local.

I have enabled simpletest on the PR sandbox and run the many to one test, and there were `2 fails, 0 exceptions, and 6 debug messages`:

http://2768.backdrop.backdrop.qa.backdropcms.org/admin/config/development/testing/results/1

<img width="1132" alt="Screen Shot 2019-06-30 at 7 43 19 pm" src="https://user-images.githubusercontent.com/2423362/60394946-574f4a00-9b6f-11e9-828d-88c6102f861a.png">

**Comment by docwilmot** [comment]
So its a problem with Simpletest vs Zen. Thats a new one, and a bit strange. I'd say though the code works, the tests pass, so commit anyway? I guess @quicksketch or @serundeputy or @herbdool will need to decide on this.

**Comment by jenlampton** [comment]
> Thought I'd try the original test from the D7 patch

The D7 test compares node IDs to user IDs, and since we have default content, our node IDs are different. Just because the tests pass, that doesn't mean it's working. There needs to be some serious changes to the test so that it will test the same thing on Backdrop that it was testing in Drupal...  hence the hours and hours of work...

**Comment by docwilmot** [comment]
But a test that passes on local shouldnt fail on Zen, using the same codebase.

**Comment by jenlampton** [comment]
Update based on today's issue, [this is the commit](https://git.drupalcode.org/project/views/commit/fbafd79a763f19d8c06039f3416a1890d5bfe143) in Drupal that caused the breakage. It's a security fix, so cannot simply be reverted.

**Comment by jenlampton** [comment]
I wonder if it is the long table names setting again that causes tests to fail on Zen?

**Comment by jenlampton** [comment]
I kept checking this issue for an update about the views that worked/didn't work, but since there was no comment, I had assumed it had been overlooked. I just now saw that id had been posted in Gitter... Moving it here for people who work on issues, in issues (ahem, me!)

> just checked locally, the view i used to test @jenlampton PR from https://github.com/backdrop/backdrop-issues/issues/3614#issuecomment-506962644 was deleted, but the step by step is in that comment. @jenlampton using your PR, does a view work when you filter by multiple NOTs? cause thats what didn't work for me using your PR.

**Comment by docwilmot** [comment]
Apologies for posting issues where issues dont belong. @jenlampton and anyone else though, can we clarify this so we can try and fix it please? 

https://github.com/backdrop/backdrop/pull/2754 by @jenlampton uses [this patch](https://www.drupal.org/files/issues/2019-03-25/views-not-operator-3040391-25.patch) from Drupal
https://github.com/backdrop/backdrop/pull/2768 by @docwilmot uses [this patch](https://www.drupal.org/project/views/issues/3040391#comment-13052907) from Drupal; this was the one committed to D7.

Both use the same tests (not sure why https://github.com/backdrop/backdrop-issues/issues/3614#issue-424447236 says the PR I did doesnt test the right things). except the test in @jenlampton's PR deletes core nodes before running the rest of the test.

@jenlampton patch didnt work for me when used on multiple NOTs; it would be good if someone could please recheck that this is the case.

The PR by @docwilmot  passes all tests locally but fails on Zen if this block is commented out:
```
    $this->assertEqual($result_count, 1, 'View has one result.');
    $nid = isset($view->result[0]->nid) ? (int) $view->result[0]->nid : 0;
    $this->assertIdentical($nid, (int) $this->nodes[1]->nid, 'View result has correct node id.');
```
but if that block is left in, it passes on Zen but fails locally (and although the test on the sandbox is passing, if the same sandbox runs the test at `development/testing` it fails on the sandbox curiously, so it seems to be a specifically Simpletest fail)

Thats the summary, to be clear.

It would be great if we could test both approaches again.

So if @jenlampton patch passes for some people but not others (multiple NOTs or not) then probably its a platform specific issue.

And I have no explanation for the fail on mine.

Hopefully we can move forward after this.

**Comment by docwilmot** [comment]
I exported the [View used in the test](https://github.com/backdrop/backdrop/pull/2768/files#diff-8f99ecc942b6d264e766cf5ed6821711R331) using [Default Views Convert to Config](https://github.com/backdrop-contrib/default_views_config) module then imported it to the sandbox site (see http://2768.backdrop.backdrop.qa.backdropcms.org/admin/structure/views/view/test_oneof) and also to a local install.

As noted https://github.com/backdrop/backdrop/pull/2768 passes all tests.  On my local, that particular test still fails. 

BUT, on my local, in the test, instead of just reading Views results from `executeView()` like the test does, I added a `$view->save()` and a call to `$this->backdropGet('admin/structure/views/view/test_oneof ')` to see the UI output. Strangely, the View *does* have the expected result in the UI. There is the expected node title there. but somehow in the test `$this->executeView($view);` and then checking `$view->result` gives an empty array.

I rebuilt the test as an actual View on my local, added all the fields it requires, created the users, then imprted the View from `Default Views Convert to Config`. Again it works in the UI as expected.

So in summary, the PR https://github.com/backdrop/backdrop/pull/2768 
- passes on Zen CI
- it fails in local Simpletest, but when the test UI is visually inspected, the correct output is there
- when used in an actual site, it works as expected.

I conclude that somehow, on local, `$this->executeView($view);` gives an empty result, but this is a failing of `executeView()`.

**Comment by jenlampton** [comment]
@docwilmot Does this look like the view you created when you tried to reproduce the result manually? https://gist.github.com/jenlampton/519578ba79a3605b8fb32aaacc953647

I'm seeing the same thing as you described: my manual result matches the expected result in the test, but the test still fails.

**Comment by docwilmot** [comment]
@jenlampton no the `getGroupedOneOfTestView` fails for me, on my PR. The `getReducedExposedGroupedOptionsTestView` passes both local and Zen.

**Comment by jenlampton** [comment]
I'm running these tests against Backdrop before we broke views - so I'm testing the tests, not the fix. They still don't pass, so It's definitely the tests that are the problem.

**Comment by jenlampton** [comment]
Well, I again spent way too many hours trying to figure out why tests fail when repeating the test manually in the UI works perfectly -- but I think it might have something to do with the other missing patches from Views. SO...  I have done the thing we talked about in the last few meetings, and added all the tests from Drupal but commented out those that are failing.

My hope is that once we get all the other fixes from views merged in we can uncomment those and they will pass.

**Comment by quicksketch** [comment]
I'm hesitant to commit anything with the intention to fix the tests later, however... the current PR fixes a known issue and has been confirmed through manual testing AND it does expand test coverage beyond what we have currently.

I think it unlikely the remaining tests will be fixed in the near future. Writing/fixing tests is tedious work and if we don't write them at the time we're fixing bugs or adding features, it's unlikely they'll get fixed at all. I appreciate #4092 and #4049 so that we don't entirely loose track of those issues.

So all that is to say I've merged https://github.com/backdrop/backdrop/pull/2878 into 1.x and 1.14.x. Thanks everyone here for their extensive testing, several attempts at fixing the problems, and just sifting through all the different comments and confusion.

Thanks everyone!
